File: sample-java/src/main/java/com/what3words/androidwrappersamplejava/MainActivity.java
package com.what3words.androidwrappersamplejava;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.text.TextUtils;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.PermissionChecker;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;
import com.what3words.androidwrapper.What3WordsV3;
import com.what3words.androidwrapper.voice.VoiceBuilder;
import com.what3words.javawrapper.request.Coordinates;
import com.what3words.javawrapper.response.Suggestion;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class MainActivity extends AppCompatActivity {

    private VoiceBuilder voiceBuilder = null;
    MaterialButton buttonAutoSuggestVoice;
    TextView volumeAutoSuggestVoice;
    TextView resultAutoSuggestVoice;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        What3WordsV3 wrapper = new What3WordsV3("YOUR_API_KEY_HERE", this);

        MaterialButton buttonConvertTo3wa = findViewById(R.id.buttonConvertTo3wa);
        TextInputEditText textInputConvertTo3wa = findViewById(R.id.textInputConvertTo3wa);
        TextView resultConvertTo3wa = findViewById(R.id.resultConvertTo3wa);

        //convert-to-3wa sample
        buttonConvertTo3wa.setOnClickListener(view -> {
            try {
                String[] latLong = textInputConvertTo3wa.getText().toString().replaceAll("\\s", "").split(",");
                Double lat = Double.parseDouble(latLong[0]);
                Double lng = Double.parseDouble(latLong[1]);
                if (lat != null && lng != null) {
                    Observable.fromCallable(() -> wrapper.convertTo3wa(new Coordinates(lat, lng)).execute())
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe(result -> {
                                if (result.isSuccessful()) {
                                    resultConvertTo3wa.setText(String.format("3 word address: %s", result.getWords()));
                                } else {
                                    resultConvertTo3wa.setText(result.getError().getMessage());
                                }
                            });
                } else {
                    resultConvertTo3wa.setText("invalid lat,long");
                }
            } catch (Exception e) {
                resultConvertTo3wa.setText("invalid lat,long");
            }
        });

        MaterialButton buttonConvertToCoordinates = findViewById(R.id.buttonConvertToCoordinates);
        TextInputEditText textInputConvertToCoordinates = findViewById(R.id.textInputConvertToCoordinates);
        TextView resultConvertToCoordinates = findViewById(R.id.resultConvertToCoordinates);

        //convert-to-coordinates sample
        buttonConvertToCoordinates.setOnClickListener(view -> {
            String address = "";
            if (textInputConvertToCoordinates.getText() != null)
                address = textInputConvertToCoordinates.getText().toString();
            String finalAddress = address;
            Observable.fromCallable(() -> wrapper.convertToCoordinates(finalAddress).execute())
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(result -> {
                        if (result.isSuccessful()) {
                            resultConvertToCoordinates.setText(String.format("Coordinates: %s,%s", result.getCoordinates().getLat(), result.getCoordinates().getLng()));
                        } else {
                            resultConvertToCoordinates.setText(result.getError().getMessage());
                        }
                    });

        });

        MaterialButton buttonAutoSuggest = findViewById(R.id.buttonAutoSuggest);
        TextInputEditText textInputAutoSuggest = findViewById(R.id.textInputAutoSuggest);
        TextView resultAutoSuggest = findViewById(R.id.resultAutoSuggest);

        //autosuggest sample
        buttonAutoSuggest.setOnClickListener(view -> {
            String query = "";
            if (textInputAutoSuggest.getText() != null)
                query = textInputAutoSuggest.getText().toString();
            String finalQuery = query;
            Observable.fromCallable(() -> wrapper.autosuggest(finalQuery).execute())
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(result -> {
                        if (result.isSuccessful()) {
                            List<String> suggestionsWords = new ArrayList<>();
                            for (Suggestion suggestion : result.getSuggestions()) {
                                suggestionsWords.add(suggestion.getWords());
                            }
                            resultAutoSuggest.setText(String.format("Suggestions: %s", TextUtils.join(",", suggestionsWords)));
                        } else {
                            resultAutoSuggest.setText(result.getError().getMessage());
                        }
                    });

        });

        buttonAutoSuggestVoice = findViewById(R.id.buttonAutoSuggestVoice);
        volumeAutoSuggestVoice = findViewById(R.id.volumeAutoSuggestVoice);
        resultAutoSuggestVoice = findViewById(R.id.resultAutoSuggestVoice);

        //voice autosuggest sample
        VoiceBuilder.Microphone microphone = new VoiceBuilder.Microphone().onListening(volume -> {
            if (volume != null) {
                volumeAutoSuggestVoice.setText(String.format("volume: %s", Math.round(volume * 100)));
            }
        });

        voiceBuilder = wrapper.autosuggest(microphone, "en")
                .onSuggestions(suggestions -> {
                    List<String> suggestionsWords = new ArrayList<>();
                    for (Suggestion suggestion : suggestions) {
                        suggestionsWords.add(suggestion.getWords());
                    }
                    resultAutoSuggestVoice.setText(String.format("Suggestions: %s", TextUtils.join(",", suggestionsWords)));
                    buttonAutoSuggestVoice.setIconResource(R.drawable.ic_record);
                })
                .onError(error -> {
                    buttonAutoSuggestVoice.setText(error);
                    buttonAutoSuggestVoice.setIconResource(R.drawable.ic_record);
                });

        buttonAutoSuggestVoice.setOnClickListener(view -> {
            if (voiceBuilder != null && voiceBuilder.isListening()) {
                buttonAutoSuggestVoice.setIconResource(R.drawable.ic_record);
                voiceBuilder.stopListening();
            } else if (voiceBuilder != null && !voiceBuilder.isListening()) {
                int permission = PermissionChecker.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO);
                if (permission == PermissionChecker.PERMISSION_GRANTED) {
                    buttonAutoSuggestVoice.setIconResource(R.drawable.ic_stop);
                    voiceBuilder.startListening();
                } else {
                    ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.RECORD_AUDIO}, 1);
                }
            }
        });
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == 1 && grantResults.length == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            buttonAutoSuggestVoice.setIconResource(R.drawable.ic_stop);
            voiceBuilder.startListening();
        } else {
            resultAutoSuggestVoice.setText("record audio permission required");
        }
    }
}

File: sample/src/main/java/com/what3words/androidwrappersample/MainActivity.kt
package com.what3words.androidwrappersample

import android.Manifest
import android.annotation.SuppressLint
import android.content.pm.PackageManager
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.PermissionChecker
import com.what3words.androidwrapper.What3WordsV3
import com.what3words.androidwrapper.voice.Microphone
import com.what3words.androidwrapper.voice.VoiceBuilder
import com.what3words.androidwrappersample.databinding.ActivityMainBinding
import com.what3words.androidwrappersample.databinding.ActivityMainBinding.inflate
import com.what3words.javawrapper.request.Coordinates
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlin.math.roundToInt

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private var builder: VoiceBuilder? = null
    private val wrapper by lazy {
        What3WordsV3("YOUR_API_KEY_HERE", this)
    }

    @SuppressLint("SetTextI18n")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = inflate(layoutInflater)

        // convert-to-3wa sample
        binding.buttonConvertTo3wa.setOnClickListener {
            val latLong = binding.textInputConvertTo3wa.text?.replace("\\s".toRegex(), "")?.split(",")
                ?.filter { it.isNotEmpty() }
            val lat = latLong?.getOrNull(0)?.toDoubleOrNull()
            val long = latLong?.getOrNull(1)?.toDoubleOrNull()
            if (lat != null && long != null) {
                CoroutineScope(Dispatchers.IO).launch {
                    // use wrapper.convertTo3wa() with Dispatcher.IO - background thread
                    val result = wrapper.convertTo3wa(Coordinates(lat, long)).execute()
                    CoroutineScope(Dispatchers.Main).launch {
                        // use Dispatcher.Main to update your views with the results - Main thread
                        if (result.isSuccessful) {
                            binding.resultConvertTo3wa.text = "3 word address: ${result.words}"
                        } else {
                            binding.resultConvertTo3wa.text = result.error.message
                        }
                    }
                }
            } else {
                binding.resultConvertTo3wa.text = "invalid lat,long"
            }
        }

        // convert-to-coordinates sample
        binding.buttonConvertToCoordinates.setOnClickListener {
            CoroutineScope(Dispatchers.IO).launch {
                // use wrapper.convertToCoordinates() with Dispatcher.IO - background thread
                val result =
                    wrapper.convertToCoordinates(binding.textInputConvertToCoordinates.text.toString())
                        .execute()
                CoroutineScope(Dispatchers.Main).launch {
                    // use Dispatcher.Main to update your views with the results - Main thread
                    if (result.isSuccessful) {
                        binding.resultConvertToCoordinates.text =
                            "Coordinates: ${result.coordinates.lat}, ${result.coordinates.lng}"
                    } else {
                        binding.resultConvertToCoordinates.text = result.error.message
                    }
                }
            }
        }

        // text autosuggest sample
        binding.buttonAutoSuggest.setOnClickListener {
            CoroutineScope(Dispatchers.IO).launch {
                // use wrapper.autosuggest() with Dispatcher.IO - background thread
                val result = wrapper.autosuggest(binding.textInputAutoSuggest.text.toString()).execute()
                CoroutineScope(Dispatchers.Main).launch {
                    // use Dispatcher.Main to update your views with the results - Main thread
                    if (result.isSuccessful) {
                        binding.resultAutoSuggest.text = if (result.suggestions.count() != 0)
                            "Suggestions: ${result.suggestions.joinToString { it.words }}"
                        else "No suggestions available"
                    } else {
                        binding.resultAutoSuggest.text = result.error.message
                    }
                }
            }
        }

        val microphone = Microphone().onListening {
            it?.let { volume ->
                binding.volumeAutoSuggestVoice.text =
                    "volume: ${(volume.times(100).roundToInt())}"
            }
        }.onError {
            binding.buttonAutoSuggestVoice.setIconResource(R.drawable.ic_record)
            binding.resultAutoSuggestVoice.text = it
        }

        // voice autosuggest sample
        builder = wrapper.autosuggest(microphone, "en")
            .focus(Coordinates(51.457269, -0.074788))
            .onSuggestions { suggestions ->
                binding.buttonAutoSuggestVoice.setIconResource(R.drawable.ic_record)
                binding.resultAutoSuggestVoice.text =
                    "Suggestions: ${suggestions.joinToString { it.words }}"
            }.onError { error ->
                binding.buttonAutoSuggestVoice.setIconResource(R.drawable.ic_record)
                binding.resultAutoSuggestVoice.text = "${error.key}, ${error.message}"
            }

        binding.buttonAutoSuggestVoice.setOnClickListener {
            if (builder?.isListening() == true) {
                binding.buttonAutoSuggestVoice.setIconResource(R.drawable.ic_record)
                builder?.stopListening()
            } else {
                // Check if RECORD_AUDIO permission is granted
                val permission =
                    PermissionChecker.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO)

                if (permission == PermissionChecker.PERMISSION_GRANTED) {
                    binding.buttonAutoSuggestVoice.setIconResource(R.drawable.ic_stop)
                    builder?.startListening()
                } else {
                    // request RECORD_AUDIO permission
                    ActivityCompat.requestPermissions(
                        this,
                        arrayOf(Manifest.permission.RECORD_AUDIO),
                        1
                    )
                }
            }
        }

        setContentView(binding.root)
    }

    @SuppressLint("SetTextI18n")
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == 1 && grantResults.count() == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            binding.buttonAutoSuggestVoice.setIconResource(R.drawable.ic_stop)
            builder?.startListening()
        } else {
            binding.resultAutoSuggestVoice.text = "record audio permission required"
        }
    }
}
